### ISSUE TITLE
App Service Graceful Shutdown / Migrate Connections help

### ISSUE THREAD

**Issue by andrew-tevent** [Azure/azure-signalr#1265]
Framework: .net5.0 
Microsoft.Azure.SignalR: v1.8.0 

I have an Azure App Service which I am using as the "Hub Server" in "Default Mode" and I deploy to a staging slot on the service to then swap in the new version.

I have found the documentation on [Graceful Shutdown](https://docs.microsoft.com/en-us/azure/azure-signalr/server-graceful-shutdown) and also read the settings documentation around [GracefulShutdown.Mode](https://github.com/Azure/azure-signalr/blob/16029cbde8f45af6fbac2923c5b7102e34d5a0c9/docs/run-asp-net-core.md#gracefulshutdown).

My deployment Pipeline deploys the new version to the staging slot then swaps it in to the production.   At this point my clients may be "connected" (proxied through the Azure SignalR service) to the *old* version in the staging slot.

I've configured, as per the documentation, the Graceful Shutdown to Migrate the connections over:

```
options.GracefulShutdown.Mode = GracefulShutdownMode.MigrateClients;
options.GracefulShutdown.Timeout = TimeSpan.FromSeconds(90);
options.GracefulShutdown.Add<Hubs.NotificationHub>(c =>
{
    Trace.WriteLine("GRACEFUL SHUTDOWN");
    return c.Clients.All.SendAsync("GracefulShutdown");
});
```

And after the swap my Pipeline *stops* the staging slot. 

But from what I can see - there is no migration, or even graceful shutdown - the client connections get disconnected immediately and reconnect (sometimes the Disconnect is actually logged _after_ the Reconnect, but this is a side issue).

Is there something I am doing wrong?  Is this the correct approach?

If I don't shut down the staging slot then clients remain connected to the old version in the staging slot and new clients can even connect to it.   In the [documentation](https://docs.microsoft.com/en-us/azure/azure-signalr/server-graceful-shutdown) it talks about "Set the Server Offline" - but it isn't clear about the approach for this.  Should I be adding a sticky slot setting to do this - something like removing the Azure SignalR connection string?  But as I understand on the swap, this would cause a settings change and an application reload therefore terminating the connections anyway?

Could someone give me some guidance on this?  Firstly why I see no appearance of the Graceful Shutdown, and then any advice on the recommended way to do this?

**Comment by terencefan** [comment]
I'll see if it works well, sorry for any inconvenience.

**Comment by andrew-tevent** [comment]
I should have mentioned that I added the checks for the migration feature into the Hub OnConnectedAsync/DisconnectedAsync:

```
var httpContext = Context.GetHttpContext();
var migration = httpContext.Features.Get<IConnectionMigrationFeature>();
if (migration != null)
       _logger.LogWarning($"OnConnectedAsync: [{migration.MigrateTo}] {Context.ConnectionId} is migrated from {migration.MigrateFrom}.");
```

But nothing gets logged.

**Comment by andrew-tevent** [comment]
@terencefan Did you validate that this is working.

In the last couple of days I've started to see Connections with the *same* ConnectionID as previous connections come into the OnConnectedAsync.  These are possibly related to server events, but I can't be sure.  But still, the `IConnectionMigrationFeature` is returning `null` so I can't tell whether this is related to migration or not....

**Comment by andrew-tevent** [comment]
Example of same Connection ID being used - see highlighted

![image](https://user-images.githubusercontent.com/78795042/122553409-89456980-d02f-11eb-9276-dac804950685.png)

**Comment by andrew-tevent** [comment]
I've upgraded to Microsoft.Azure.SignalR: v1.8.2 but still seeing issues:

1. `IConnectionMigrationFeature` is always `null` so never informs me of anything being migrated.
2.  On shutting down my staging environment it now triggers a Disconnect and Connect (not always in that order, mind you) - presumably with the **same** connection ID _by design_ BUT, the client is then additionally disconnected and has to reconnect (i.e. another Disconnect and Connect presumably caused by the connection still being interrupted between the Server and Azure SignalR?)

**Comment by andrew-tevent** [comment]
Is there any chance of an update on this?

I've had to turn off the Migration stuff because it doesn't seem reliable or to be working correctly.

I'm now fighting with an issue with `OnDisconnectAsync` not appearing to be called, possibly during a deployment of a new version of my server - which may or may not be due to me shutting down the staging environment?

**Comment by terencefan** [comment]
https://github.com/Azure/azure-signalr/pull/1376
The bug is fixed in `1.10.0`.
It was broke by one of our latest changes that allowed the application task to stand alone for a while after the connection has been closed.
